services/offline/src/lib/dhis2-connection-status.tsx

```diff
@@ -0,0 +1,221 @@
+import { useConfig } from '@dhis2/app-service-config'
+import PropTypes from 'prop-types'
+import React, {
+    useCallback,
+    useState,
+    useRef,
+    useMemo,
+    useEffect,
+    useContext,
+} from 'react'
+import { useOfflineInterface } from './offline-interface'
+import createSmartInterval, { SmartInterval } from './smart-interval'
+import { usePingQuery } from './use-ping-query'
+
+// Utils for saving 'last connected' datetime in local storage
+const lastConnectedKey = 'dhis2.lastConnected'
+type AppName = string | undefined
+export const getLastConnectedKey = (appName?: AppName) =>
+    appName ? `${lastConnectedKey}.${appName}` : lastConnectedKey
+const updateLastConnected = (appName: AppName) => {
+    // use Date.now() because it's easier to mock for easier unit testing
+    const now = new Date(Date.now())
+    localStorage.setItem(getLastConnectedKey(appName), now.toUTCString())
+    return now
+}
+const getLastConnected = (appName: AppName) => {
+    const lastConnected = localStorage.getItem(getLastConnectedKey(appName))
+    return lastConnected ? new Date(lastConnected) : null
+}
+const clearLastConnected = (appName: AppName) => {
+    localStorage.removeItem(getLastConnectedKey(appName))
+}
+
+export interface Dhis2ConnectionStatus {
+    isConnected: boolean
+    isDisconnected: boolean
+    lastConnected: Date | null
+}
+
+const Dhis2ConnectionStatusContext = React.createContext({
+    isConnected: true,
+    isDisconnected: false,
+    lastConnected: null,
+} as Dhis2ConnectionStatus)
+
+/**
+ * Provides a boolean indicating client's connection to the DHIS2 server,
+ * which is different from connection to the internet.
+ *
+ * The context provider subscribes to messages from the SW tracking successes
+ * and failures of requests to the DHIS2 server to determine connection status,
+ * and then will initiate periodic pings if there are no incidental requests in
+ * order to check the connection consistently
+ */
+export const Dhis2ConnectionStatusProvider = ({
+    children,
+}: {
+    children: React.ReactNode
+}): JSX.Element => {
+    const offlineInterface = useOfflineInterface()
+    const { appName } = useConfig()
+    // The offline interface persists the latest update from the SW so that
+    // this hook can initialize to an accurate value. The App Adapter in the
+    // platform waits for this value to be populated before rendering the
+    // the App Runtime provider (including this), but if that is not done,
+    // `latestIsConnected` may be `null` depending on the outcome of race
+    // conditions between the SW and the React component tree.
+    const [isConnected, setIsConnected] = useState(
+        offlineInterface.latestIsConnected
+    )
```

Comment:
Implemented in this commit: https://github.com/dhis2/app-platform/pull/718/commits/5644f0ddaa428a2dcbd6ed86e2004299cfb11a02